src/audio/peak_volume.c

```diff
@@ -0,0 +1,335 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component_ext.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/volume.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/ipc/msg.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/drivers/timer.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/cache.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/notifier.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc4/module.h>
+#include <ipc4/peak_vol.h>
+#include <ipc/dai.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* Purpose of Peak Volume/Meter (PeakVol) is to:
+ * 1. Measure input volume (amplitude)
+ * 2. Change signal volume (optional)
+ * 3. Fade signal	(optional)
+ * Fading signal consumes a lot of MCPS. It is recommended
+ * to use signal fading only for playback purposes.
```

Comment:
@singalsu Whats the feature overlap we have today with volume ? What do we need to do and whats missing ?

Reply:
We don't have have "1. Measure input volume (amplitude)". Features 2. and 3. exist.  But also I can't find from this patch nor topic/ipc4 any level measurement code for feature 1. from processing done copy().  Is this patch work-in-progress?

Reply:
Upstream volume also has ZC which I dont think is requirement for OS but is for codec HW.

Reply:
@singalsu yes, we need to add more features. Can you help to check peak_vol_module.cc whether we fully support the curve algorithm? Thanks!